### PR TITLE
Data generator

### PR DESCRIPTION
**Pull request by l8git**
- changed output
- renaming of self.attr to self._attr
- adding device to use GPU when available

### CODE REVIEW

File: data_generator.py [hunk 1]
```diff
@@ -85,68 +88,68 @@ def _kernel(self, X, length, sigma, noise=2e-2):
...
-        num_context = torch.randint(3, self.max_num_context, [])
+        num_context = torch.randint(3, self._max_num_context, [])
 
-        if self.testing:
-            num_target = 401
+        if self._testing:
+            num_target = 400
             num_total_points = num_target
-            X = torch.range(-2, 2, 1. / 100).unsqueeze(0).expand(self.batch_size, -1)
+            X = torch.arange(-2, 2, 0.01).unsqueeze(0).expand(self._batch_size, -1)
```
Comment (rodrigobdz) [inline data_generator.py:111]: @l8git What's the advantage here of `arange` over `range`?

File: data_generator.py [hunk 2]
```diff
@@ -85,68 +88,68 @@ def _kernel(self, X, length, sigma, noise=2e-2):
...
+        # [batch_size, y_size, num_total_points, num_total_points]
         kernel = self._kernel(X, length, sigma)
-        # TODO: (maybe) change precision to float64 and cast to float32 afterwards
+
+        # change precision to float64 for Cholesky and cast to float32 afterwards
         cholesky = kernel.double().cholesky().float()
-        y = cholesky.matmul(torch.randn(self.batch_size, self.y_size, num_total_points, 1))
+
         # sampling with no mean assumption: y = mu + sigma*z~N(0,I) ~ c.L * rand_normal([0, 1]) with appropriate shape
-        # TODO: if runtime error: change dimension -1 of torch.randn_like(cholesky) to ?
```
Comment (rodrigobdz) [inline data_generator.py:136]: @l8git The var `y` seems to be assigned the same value as before.
Is this `TODO` already solved? If yes, how?

File: data_generator.py [hunk 3]
```diff
@@ -157,7 +160,11 @@ def generate_curves(self):
...
 
-        query = ((context_x, context_y), target_x)
+        context_x = context_x.to(self._dev)
+        context_y = context_y.to(self._dev)
+        target_x = target_x.to(self._dev)
+        target_y = target_y.to(self._dev)
+
 
-        return query, target_y
+        return (context_x, context_y, target_x, target_y)
```
Comment (rodrigobdz) [inline data_generator.py:169]: @l8git Is `query` not needed anymore?

File: data_generator.py (hunk 1 above)
Comment (l8git) [reply data_generator.py:111]: no advantage, just deprecated

File: data_generator.py (hunk 3 above)
Comment (l8git) [reply data_generator.py:169]: no, only the tuple will be used and unpacked when needed

File: data_generator.py (hunk 2 above)
Comment (l8git) [reply data_generator.py:136]: regarding `y`: yes, it just happens in a different row now

regarding the `TODO`: my bad before, just usual matrix multiplication that works

**Pull request merged**